### DISCUSSION TITLE
Profiling micropython running on Lego machine

### DISCUSSION THREAD

**Discussion by jaguilar** [pybricks/support#1441] [Q&A]
I have a program I'd like to make run faster, just for my own interest. At work, I typically fire up a profiler for something like this, but I'm not sure if any such exists on pybricks. Do we have anything like that?

**Comment by laurensvalk** [comment]
There is the StopWatch class which would let you time things. There is also micropython.mem_info to get stats on stack and heap usage along the way.

**Reply by jaguilar** [reply]
Thanks! I should have been a little more specific. I'm thinking more like function or line level cycle count information a la [`profile`](https://docs.python.org/3/library/profile.html#module-profile). 

The StopWatch can tell me how long things take on a gross scale, but if my PID controller is slower than I would like, it can't easily tell me which part is slow. I thought it might be helpful to get more granular information to help guide my optimization efforts.

Alternatively, if there is a general performance guide somewhere, that might also be helpful.
